src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -274,13 +280,17 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
                 // if we can't allocate it on a node, ignore it, for example, this handles
                 // cases for only allocating a replica after a primary
                 Decision decision = allocation.deciders().canAllocate(shard, node, allocation);
+                if (decision.type() != Decision.Type.THROTTLE) {
```

Comment:
actually, yea, I think you are right, it makes this optimization invalid, argh, I already thought about it before :)